src/scenes/Engagement/EditEngagement/EditEngagementDialog.tsx

```diff
@@ -26,186 +25,172 @@ import {
 } from '../../../components/form';
 import { CountryField, UserField } from '../../../components/form/Lookup';
 import { UserLookupItemFragment } from '../../../components/form/Lookup/User/UserLookup.generated';
-import { InternshipEngagementDetailFragment } from '../InternshipEngagement/InternshipEngagement.generated';
-import { LanguageEngagementDetailFragment } from '../LanguageEngagement/LanguageEngagementDetail.generated';
+import { many, Many } from '../../../util';
+import { InternshipEngagementDetailFragment as InternshipEngagement } from '../InternshipEngagement/InternshipEngagement.generated';
+import { LanguageEngagementDetailFragment as LanguageEngagement } from '../LanguageEngagement/LanguageEngagementDetail.generated';
 import {
   useUpdateInternshipEngagementMutation,
   useUpdateLanguageEngagementMutation,
 } from './EditEngagementDialog.generated';
 
-type EditEngagementDialogProps = Except<
-  DialogFormProps<
-    UpdateInternshipEngagementInput | UpdateInternshipEngagementInput
-  >,
-  'onSubmit' | 'initialValues'
-> & {
-  engagement:
-    | InternshipEngagementDetailFragment
-    | LanguageEngagementDetailFragment;
-  editValue?: string;
+type Engagement = InternshipEngagement | LanguageEngagement;
+type EngagementKeys = keyof InternshipEngagement | keyof LanguageEngagement;
+
+interface EngagementFieldProps {
+  engagement: Engagement;
+}
+
+const fieldMapping = {
```

Comment:
Okay, so this is very pretty and satisfying, and the nested ternary operators below are messy and feel brittle, but if the DX is so bad on this, could you instead keep the ternaries and just create yourself an Enum instead of all the magic strings? That'd be a compromise and wouldn't feel as nice, but **would** still provide the missing strictness you're looking for.

If you're fine with the DX, then otherwise this looks good to me.

Reply:
The code is strict. I think the DX is bad because the validation isn't on the key/value pair but on an unused variable below. But on the other hand the DX for the object mapping in general is way better than the ternaries IMO.
Let me try duplicating the keys in a string literal and see how that looks.

Reply:
Ok pushed. I'm more happy with that anyways. Thanks for the push (ha) 👍 Lemme know what you think.